docs/20-design.md

```diff
@@ -0,0 +1,59 @@
+
+For more on this topic, see Casey Bisson's article on the ContainerSummit website: [_Active vs. passive discovery in distributed applications_](https://containersummit.io/articles/active-vs-passive-discovery).
+
+
+## Why isn't there a "post-start" or "started" event?
+
+In theory it's possible for ContainerPilot to fire an event immediately after it `fork/exec`'s a job's process. But in practice this is not a useful event for an application developer and will cause hard-to-debug race conditions in your configuration. Processes like servers need to bind to an address before they can be considered "started." Other processes need to do other work during start like loading and interpreting byte code. A short-lived task might even complete all its work and exit before an event could be handled, leading to an unsolvable race condition.
+
+If ContainerPilot were to emit a `started` event then the job that emits the event will almost never be ready to do work when the event is handled. This means that jobs that depend on that job will need to implement some kind of health checking and retry logic anyways, and we would gain nothing at the cost of making configuration very confusing and error prone. Jobs that are dependent on another job should listen for the `healthy` event from their dependencies.
+
+
+## Why Consul and not etcd or Zookeeper?
+
+Consul provides a number of higher-level capabilities than a simple KV store like etcd or ZK. Providing the ability to use these capabilities would mean either going with a least-common-denominator approach or having complex provider-specific configuration options for tagging interfaces, providing secure connection to the backend, and faster deregistration on shutdown, among others. Additionally, Consul has first-class support for multi-datacenter deployments.
+
+The primary argument for supporting etcd rather than Consul is that Kubernetes and related projects are using it as their service discovery layer. In our discussions with some end users, we haven't found that there's any resistance to the idea that the scheduler's own consensus & membership store doesn't need to be the same store used by applications. And even perhaps _should not_ be the same store, given that in most organizations the team responsible for application development will not be the same team responsible for running the deployment platform.
+
+
+## Why are jobs not the same as services?
+
+A job is the core abstraction for managing all processes in ContainerPilot. But the end user will not necessarily want to advertise all processes in a container to Consul. Health checks, sensors, setup tasks, etc. are all processes that a container needs to run that are internal to the container.
+
+
+## Why don't watches or metrics have behaviors?
+
+By having ContainerPilot's behaviors all in jobs rather than in watches or metrics, you can configure more than one behavior when an event is emitted.
```

Comment:
Updated to:

> ## Why don't watches or metrics have an exec field?

> By having ContainerPilot's `exec` fields all in jobs rather than in watches or metrics, you can configure more than one `exec` behavior when an event is emitted.